core/arch/arm/kernel/spmc_sp_handler.c

```diff
@@ -216,10 +216,87 @@ static int spmc_sp_add_sp_region(struct sp_mem *smem,
 	return res;
 }
 
+static int release_mobj(struct sp_mem *smem, struct sp_mem_map_region *region,
+			bool mobj_ffa)
+{
+	uint64_t handle = smem->transaction.global_handle;
+	int res = FFA_OK;
+
+	if (mobj_ffa) {
+		if (handle) {
+			mobj_put(region->mobj);
+			if (mobj_ffa_unregister_by_cookie(handle))
+				return FFA_DENIED;
+
+			if (mobj_ffa_sel1_spmc_reclaim(handle))
+				return FFA_DENIED;
+		} else {
+			mobj_ffa_sel1_spmc_delete(to_mobj_ffa(region->mobj));
+		}
+	} else {
+		mobj_put(region->mobj);
+	}
+
+	return res;
+}
+
 static int spmc_sp_add_nw_region(struct sp_mem *smem,
 				 struct ffa_mem_region *mem_reg)
 {
+	uint64_t page_count = READ_ONCE(mem_reg->total_page_count);
+	struct mobj *mobj = NULL;
+	struct sp_mem_map_region *region = NULL;
+	struct mobj_ffa *m = mobj_ffa_sel1_spmc_new(page_count);
+	unsigned int i = 0;
+	unsigned int idx = 0;
+	int res = FFA_OK;
+	uint64_t cookie = 0;
+
+	if (!m)
+		return FFA_DENIED;
+
+	for (i = 0; i < READ_ONCE(mem_reg->address_range_count); i++) {
+		struct ffa_address_range *addr_range = NULL;
+
+		addr_range =  &mem_reg->address_range_array[i];
+		if (mobj_ffa_add_pages_at(m, &idx,
+					  READ_ONCE(addr_range->address),
+					  READ_ONCE(addr_range->page_count))) {
+			res = FFA_DENIED;
+			goto clean_up;
+		}
+	}
+	cookie = mobj_ffa_push_to_inactive(m);
```

Comment:
After this call can the cookie be used to get a mobj by any thread, so the cleanup below may very well fail to remove the mobj.

Reply:
Not used anymore